app/scripts/lib/rpc-method-middleware/handlers/wallet-createSession/handler.ts

```diff
@@ -117,55 +145,129 @@ async function walletCreateSessionHandler(
+      supportedRequiredScopes,
+      supportedOptionalScopes,
+    ]);
+
+    const allSupportedRequestedCaipChainIds = getAllScopesFromScopesObjects([
+      supportedRequiredScopes,
+      supportedOptionalScopes,
+    ]);
+
     // Fetch EVM accounts from native wallet keyring
-    // These addresses are lowercased already
     const existingEvmAddresses = hooks
       .listAccounts()
       .map((account) => account.address);
-    const supportedEthAccounts = getEthAccounts({
-      requiredScopes: supportedRequiredScopes,
-      optionalScopes: supportedOptionalScopes,
-    })
-      .map((address) => address.toLowerCase() as Hex)
-      .filter((address) => existingEvmAddresses.includes(address));
+
+    const supportedRequestedAccountAddresses =
+      allRequestedAccountAddresses.filter((accountAddress: CaipAccountId) => {
+        const {
+          address,
+          chain: { namespace },
+          chainId: caipChainId,
+        } = parseCaipAccountId(accountAddress);
+        if (namespace === KnownCaipNamespace.Eip155) {
+          return existingEvmAddresses.some((existingEvmAddress) => {
+            return isEqualCaseInsensitive(address, existingEvmAddress);
+          });
+        }
+
+        const getNonEvmAccountAddressesForChainId =
+          hooks.getNonEvmAccountAddresses(caipChainId);
+        return getNonEvmAccountAddressesForChainId.some(
+          (existingCaipAddress) => {
+            return isEqualCaseInsensitive(accountAddress, existingCaipAddress);
+          },
+        );
+      });
 
     const requestedCaip25CaveatValue = {
       requiredScopes: getInternalScopesObject(supportedRequiredScopes),
       optionalScopes: getInternalScopesObject(supportedOptionalScopes),
       isMultichainOrigin: true,
+      sessionProperties: filteredSessionProperties,
     };
 
-    const requestedCaip25CaveatValueWithSupportedEthAccounts = setEthAccounts(
-      requestedCaip25CaveatValue,
-      supportedEthAccounts,
-    );
+    const requestedCaip25CaveatValueWithSupportedAccounts =
+      setPermittedAccounts(
+        requestedCaip25CaveatValue,
+        supportedRequestedAccountAddresses,
+      );
 
-    const [grantedPermissions] = await hooks.requestPermissionsForOrigin({
-      [Caip25EndowmentPermissionName]: {
-        caveats: [
-          {
-            type: Caip25CaveatType,
-            value: requestedCaip25CaveatValueWithSupportedEthAccounts,
-          },
-        ],
+    // if `promptToCreateSolanaAccount` is true and there are no other valid scopes requested,
+    // we add a `wallet` scope to the request in order to get passed the CAIP-25 caveat validator.
+    // This is very hacky but is necessary because the solana opt-in flow breaks key assumptions
+    // of the CAIP-25 permission specification -  namely that we can have valid requests with no scopes.
+    if (allSupportedRequestedCaipChainIds.length === 0) {
+      if (promptToCreateSolanaAccount) {
```

Comment:
As far as I can tell this doesn't exist yet and requesting just the Solana scope without Solana accounts just leaves you to potentially connect an Ethereum account. Could we strip this out until we have the full feature?

Reply:
I understand the reasoning, I'm just arguing that it is simpler for that to be included in a follow-up PR to reduce the diff and complexity here

Reply:
Totally. We could have broken out the features better if we hadn't been rushing and thought it through better... 

Though part of our goal here was to allow others to build directly off of this branch and avoid long chains off of a feature branch. 

I believe they've already started building off this branch... so I could reach out and see if we can rebase on another branch off of this one if you feel strongly about it.

Reply:
I think it shouldn't be that much work to re-add this snippet in the other branch. But it makes reasoning about the changes in this PR a lot easier. Definitely worth it IMO